File: native/turboembed/bench/binding_benchmark.cpp
// SPDX-License-Identifier: Apache-2.0
// Matched native ABI reference for java/benchmarks/BindingBenchmark.java.
#include <turboembed_prepared.h>
#include <nlohmann/json.hpp>
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstring>
#include <iostream>
#include <memory>
#include <stdexcept>
#include <vector>

using Clock = std::chrono::steady_clock;
using json = nlohmann::json;
static te_error error{};
static void check(uint32_t code) {
    if (code) throw std::runtime_error(std::to_string(code) + ": " + error.message);
}
template<class T> static T descriptor() {
    T value{}; value.struct_size = sizeof(T); value.version = TE_PREPARED_VERSION; return value;
}
template<class F> static json measure(F call, int maximum, int warmup) {
    for (int i = 0; i < warmup; ++i) call();
    std::vector<int64_t> samples; samples.reserve(maximum);
    const auto start = Clock::now();
    for (int i = 0; i < maximum; ++i) {
        const auto before = Clock::now(); call(); const auto end = Clock::now();
        samples.push_back(std::chrono::duration_cast<std::chrono::nanoseconds>(end - before).count());
        if (end - start >= std::chrono::seconds(3)) break;
    }
    const double elapsed = std::chrono::duration<double>(Clock::now() - start).count();
    auto sorted = samples; std::sort(sorted.begin(), sorted.end());
    return {{"samples_ns", samples}, {"count", samples.size()}, {"elapsed_seconds", elapsed},
        {"p50_ns", sorted[(sorted.size() - 1) * 50 / 100]},
        {"p99_ns", sorted[(sorted.size() - 1) * 99 / 100]}, {"requests_per_second", samples.size() / elapsed}};
}
int main(int argc, char **argv) try {
    if (argc < 4 || argc > 5) throw std::runtime_error("usage: binding_benchmark BUNDLE BATCH SEQUENCE [gpu|cpu]");
    const int batch = std::stoi(argv[2]), sequence = std::stoi(argv[3]);
    if (batch < 1 || batch > 32 || sequence < 4 || sequence > 256) throw std::runtime_error("invalid shape");
    // GPU is the reference device; CPU must be selected explicitly.
    const std::string device = argc == 5 ? argv[4] : "gpu";
    if (device != "gpu" && device != "cpu") throw std::runtime_error("device must be gpu or cpu");
    auto opts = descriptor<te_context_options>();
    opts.device = device == "cpu" ? TE_DEVICE_OPENVINO_CPU : TE_DEVICE_OPENVINO_GPU;
    te_context *raw_context{}; check(turboembed_prepared_v1_context_create(&opts, &raw_context, &error));
    std::unique_ptr<te_context, decltype(&turboembed_prepared_v1_context_release)> context(raw_context, turboembed_prepared_v1_context_release);
    auto ci = descriptor<te_context_info>(); check(turboembed_prepared_v1_context_info(context.get(), &ci, &error));
    te_model *raw_model{}; check(turboembed_prepared_v1_model_load(context.get(), argv[1], std::strlen(argv[1]), &raw_model, &error));
    std::unique_ptr<te_model, decltype(&turboembed_prepared_v1_model_release)> model(raw_model, turboembed_prepared_v1_model_release);
    auto mi = descriptor<te_model_info>(); check(turboembed_prepared_v1_model_info(model.get(), &mi, &error));
    if (std::strcmp(mi.tokenizer_sha256, "be50c3628f2bf5bb5e3a7f17b1f74611b2561a3a27eeab05e5aa30f411572037") != 0)
        throw std::runtime_error("benchmark token IDs require pinned MiniLM tokenizer");
    auto so = descriptor<te_slot_options>(); so.batch = batch; so.sequence_length = sequence;
    te_slot *raw_slot{}; check(turboembed_prepared_v1_slot_create(model.get(), &so, &raw_slot, &error));
    std::unique_ptr<te_slot, decltype(&turboembed_prepared_v1_slot_release)> slot(raw_slot, turboembed_prepared_v1_slot_release);
    std::vector<int32_t> ids(batch * sequence), mask(ids.size()), types(ids.size());
    for (int row = 0; row < batch; ++row) {
        const int words[] = {101, 7592, 2088, 102};
        for (int i = 0; i < 4; ++i) { ids[row * sequence + i] = words[i]; mask[row * sequence + i] = 1; }
    }
    check(turboembed_prepared_v1_slot_write_tokens(slot.get(), ids.data(), mask.data(), types.data(), ids.size(), &error));
    std::vector<float> output(batch * mi.dimension), prepared(output.size());
    std::vector<te_text> texts(batch, te_text{"hello world", 11});
    auto stats = descriptor<te_slot_stats>();
    auto bridge = [&] { check(turboembed_prepared_v1_slot_stats(slot.get(), &stats, &error)); };
    auto infer = [&] {
        te_result *result{}; check(turboembed_prepared_v1_slot_execute(slot.get(), &result, &error));
        check(turboembed_prepared_v1_result_release(result, &error));
    };
    auto text = [&] {
        check(turboembed_prepared_v1_slot_write_text(slot.get(), texts.data(), texts.size(), &error));
        te_result *result{}; check(turboembed_prepared_v1_slot_execute(slot.get(), &result, &error));
        const auto read_status = turboembed_prepared_v1_result_read(result, output.data(), output.size(), &error);
        if (read_status) { const std::string message = error.message; turboembed_prepared_v1_result_release(result, &error); throw std::runtime_error(message); }
        check(turboembed_prepared_v1_result_release(result, &error));
    };
    te_result *result{}; check(turboembed_prepared_v1_slot_execute(slot.get(), &result, &error));
    check(turboembed_prepared_v1_result_read(result, prepared.data(), prepared.size(), &error));
    check(turboembed_prepared_v1_result_release(result, &error)); text();
    double maximum = 0, squared = 0;
    for (size_t i = 0; i < output.size(); ++i) {
        const double difference = std::abs(static_cast<double>(output[i]) - prepared[i]);
        if (!std::isfinite(difference)) throw std::runtime_error("nonfinite output");
        maximum = std::max(maximum, difference); squared += difference * difference;
    }
    if (maximum > 1e-6 || std::sqrt(squared / output.size()) > 1e-7) throw std::runtime_error("prepared/text parity failed");
    json records = json::array();
    for (int repeat = 0; repeat < 3; ++repeat) {
        records.push_back({{"repeat", repeat}, {"bridge", measure(bridge, 100000, 100000)},
            {"prepared", measure(infer, 3000, 500)}, {"text_host", measure(text, 3000, 500)}});
    }
    std::cout << json{{"path", "native_abi"}, {"batch", batch}, {"sequence", sequence},
        {"device", ci.device_name}, {"runtime", ci.runtime_version}, {"model", mi.model_id},
        {"revision", mi.revision}, {"tokenizer_sha256", mi.tokenizer_sha256},
        {"text", "hello world"}, {"repeats", records}}.dump() << '\n';
} catch (const std::exception &failure) { std::cerr << failure.what() << '\n'; return 1; }

File: java/benchmarks/BindingBenchmark.java
// SPDX-License-Identifier: Apache-2.0
import ai.pipestream.turboembed.*;
import ai.pipestream.turboembed.ffm.FfmTurboEmbed;
import java.lang.management.ManagementFactory;
import java.nio.FloatBuffer;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;

/** Bounded diagnostic harness, deliberately outside the consumer artifacts. */
public final class BindingBenchmark {
    private static volatile long sink;
    private static final com.sun.management.ThreadMXBean ALLOCATIONS =
        (com.sun.management.ThreadMXBean) ManagementFactory.getThreadMXBean();
    private static String measure(Runnable call, int maximum, int warmup) {
        for (int i = 0; i < warmup; i++) { call.run(); }
        long[] samples = new long[maximum]; int count = 0;
        long start = System.nanoTime();
        do {
            long before = System.nanoTime(); call.run(); long end = System.nanoTime();
            samples[count++] = end - before;
            if (end - start >= 3_000_000_000L) { break; }
        } while (count < maximum);
        double elapsed = (System.nanoTime() - start) / 1e9;
        long thread = Thread.currentThread().threadId();
        long bytesBefore = ALLOCATIONS.getThreadAllocatedBytes(thread);
        for (int i = 0; i < 100; i++) { call.run(); }
        double allocated = (ALLOCATIONS.getThreadAllocatedBytes(thread) - bytesBefore) / 100.0;
        long[] sorted = Arrays.copyOf(samples, count); Arrays.sort(sorted);
        StringBuilder json = new StringBuilder("{\"samples_ns\":[");
        for (int i = 0; i < count; i++) { if (i != 0) { json.append(','); } json.append(samples[i]); }
        return json.append("],\"count\":").append(count).append(",\"elapsed_seconds\":").append(elapsed)
            .append(",\"p50_ns\":").append(sorted[(count - 1) * 50 / 100])
            .append(",\"p99_ns\":").append(sorted[(count - 1) * 99 / 100])
            .append(",\"requests_per_second\":").append(count / elapsed)
            .append(",\"java_allocated_bytes_per_call\":").append(allocated).append('}').toString();
    }
    public static void main(String[] args) throws Exception {
        if (args.length < 5 || args.length > 6) { throw new IllegalArgumentException("SDK BUNDLE BATCH SEQUENCE OUTPUT_JSON [gpu|cpu]"); }
        if (!ALLOCATIONS.isThreadAllocatedMemorySupported()) { throw new IllegalStateException("allocation counter unavailable"); }
        ALLOCATIONS.setThreadAllocatedMemoryEnabled(true);
        int batch = Integer.parseInt(args[2]), sequence = Integer.parseInt(args[3]);
        // GPU is the reference device; CPU must be selected explicitly.
        Device device = switch (args.length == 6 ? args[5] : "gpu") {
            case "gpu" -> Device.OPENVINO_GPU;
            case "cpu" -> Device.OPENVINO_CPU;
            default -> throw new IllegalArgumentException("device must be gpu or cpu");
        };
        try (var provider = FfmTurboEmbed.open(Path.of(args[0]));
             var context = provider.context(device, 0);
             var model = context.loadModel(Path.of(args[1])); var slot = model.slot(batch, sequence)) {
            if (!model.info().tokenizerSha256().equals("be50c3628f2bf5bb5e3a7f17b1f74611b2561a3a27eeab05e5aa30f411572037")) {
                throw new IllegalArgumentException("benchmark token IDs require pinned MiniLM tokenizer");
            }
            if (sequence < 4) { throw new IllegalArgumentException("sequence must fit hello world"); }
            TokenBuffers input = slot.inputs(); int[] words = {101, 7592, 2088, 102};
            for (int row = 0; row < batch; row++) {
                for (int i = 0; i < words.length; i++) {
                    input.ids().put(row * sequence + i, words[i]); input.attentionMask().put(row * sequence + i, 1);
                }
            }
            slot.upload();
            String[] texts = new String[batch]; Arrays.fill(texts, "hello world");
            FloatBuffer output = FloatBuffer.allocate(batch * model.info().dimension());
            Runnable bridge = () -> { SlotStats stats = slot.stats(); sink = stats.executions(); };
            Runnable prepared = () -> { try (var result = slot.execute()) { sink = result.dimension(); } };
            Runnable text = () -> {
                slot.writeText(texts); output.clear();
                try (var result = slot.execute()) { result.readInto(output); }
            };
            float[] reference = new float[output.capacity()];
            try (var result = slot.execute()) { result.readInto(FloatBuffer.wrap(reference)); }
            text.run(); double maximum = 0, squared = 0;
            for (int i = 0; i < reference.length; i++) {
                double difference = Math.abs((double) reference[i] - output.get(i));
                if (!Double.isFinite(difference)) { throw new AssertionError("nonfinite output"); }
                maximum = Math.max(maximum, difference); squared += difference * difference;
            }
            if (maximum > 1e-6 || Math.sqrt(squared / reference.length) > 1e-7) { throw new AssertionError("prepared/text parity failed"); }
            StringBuilder json = new StringBuilder("{\"path\":\"java_ffm\",\"device\":\"")
                .append(context.info().name()).append("\",\"batch\":").append(batch)
                .append(",\"sequence\":").append(sequence).append(",\"text\":\"hello world\",\"jdk\":\"")
                .append(System.getProperty("java.runtime.version")).append("\",\"repeats\":[");
            for (int repeat = 0; repeat < 3; repeat++) {
                if (repeat != 0) { json.append(','); }
                json.append("{\"repeat\":").append(repeat).append(",\"bridge\":").append(measure(bridge, 100000, 100000))
                    .append(",\"prepared\":").append(measure(prepared, 3000, 500))
                    .append(",\"text_host\":").append(measure(text, 3000, 500)).append('}');
            }
            Files.writeString(Path.of(args[4]), json.append("]}\n").toString());
        }
    }
}
